### PR TITLE
System Test: Cordon and Drain Node

### PR DESCRIPTION
**Pull request by dbhatnagar-px**
This System Test covers the following steps:
1. Create a Data Service
2. Validate Storage / Resource Template of Data Service
3. Choose a node and Cordon it
4. Delete pods of Data Service from Cordoned node, and uncordon the node later on
5. Validate deleted pod comes up on a different node in the Target Cluster
6. Validate Storage / Resource Template of Data Service
7. Start a Workload on the Data Service and validate it's running

Successful Jenkins Job Run Link: https://jenkins.pwx.dev.purestorage.com/job/PDS/job/pds-branch-build-test/125

### CODE REVIEW

File: drivers/pds/lib/utils.go [hunk 1]
```diff
@@ -2300,3 +2300,59 @@ func VerifyPxPodOnNode(nodeName string, namespace string) (bool, error) {
...
+		for _, pod := range pods {
+			if pod.Spec.NodeName == nodeName {
+				podsList = append(podsList, pod)
+			}
+		}
+	}
+	if podsList != nil && len(podsList) > 0 {
+		return podsList, nil
+	}
+	return nil, errors.New("There is no pod of the given statefulset running on the given node name")
```
Comment (lsrinivas-pure) [inline drivers/pds/lib/utils.go:2357]: can you add node name in the log

File: tests/pds/dataservice_test.go [hunk 2]
```diff
@@ -627,6 +628,88 @@ var _ = Describe("{UpgradeDataServiceImage}", func() {
...
+var _ = Describe("{CordonNodeAndDeletePod}", func() {
+	JustBeforeEach(func() {
+		StartTorpedoTest("CordonNodeAndDeletePod", "Deploys a data service, cordons one selected node, deletes the pod from that node", pdsLabels, 0)
+	})
+
+	It("Cordon a Node and Delete Pods from that node", func() {
+		Step("Deploy, Validate, Cordon Node and Delete Pods, Validate new Pod, Validate Storage, Run Workload on Data Service", func() {
+			for _, ds := range params.DataServiceToTest {
+				// Run this Test on Consul only. We can enable more Data Services later from here if needed.
+				// Consul is a random choice. We can choose any other Data service as well.
```
Comment (lsrinivas-pure) [inline tests/pds/dataservice_test.go:640]: any specific reason why are we restricting for consul only? 
Comment also says it is independent

File: tests/pds/dataservice_test.go [hunk 3]
```diff
@@ -627,6 +628,88 @@ var _ = Describe("{UpgradeDataServiceImage}", func() {
...
+				// This test case is pretty much Data service independent
+				if ds.Name != consul {
+					log.Warnf("Skipping %v dataservice ", ds.Name)
+					continue
+				}
+				isDeploymentsDeleted = false
+				deployment, _, _, err = DeployandValidateDataServices(ds, tenantID, projectID)
+				log.FailOnError(err, "Error while deploying data services")
```
Comment (lsrinivas-pure) [inline tests/pds/dataservice_test.go:648]: can you add the name of dataservice in the error log

File: tests/pds/dataservice_test.go [hunk 4]
```diff
@@ -627,6 +628,88 @@ var _ = Describe("{UpgradeDataServiceImage}", func() {
...
+				log.InfoD("Running Check Node DataService Func %v ", ds.Name)
+				nodes, err := pdslib.GetNodesOfSS(*deployment.ClusterResourceName, namespace)
+				log.FailOnError(err, "Cannot fetch nodes of the running Data Service")
+				for _, node := range nodes {
+					log.InfoD("Node Name is: %s", node.Name)
```
Comment (lsrinivas-pure) [inline tests/pds/dataservice_test.go:653]: you can add one log statement like

log.InfoD("Resource %s is deployed on nodes  %s", *deployment.ClusterResourceName,nodes)
if nodes is printing full object you can create another list with only names and print it

File: tests/pds/dataservice_test.go [hunk 5]
```diff
@@ -627,6 +628,88 @@ var _ = Describe("{UpgradeDataServiceImage}", func() {
...
+				}
+				var nodeName string
+				Step("Delete Pods and Cordon a node", func() {
+					nodeName = nodes[0].Name // Selecting the 1st node in the list to cordon
+					podsList, err := pdslib.GetPodsOfSsByNode(*deployment.ClusterResourceName, nodeName, namespace)
+					log.FailOnError(err, "Pod not found on this Node")
```
Comment (lsrinivas-pure) [inline tests/pds/dataservice_test.go:659]: add node name in the log

File: tests/pds/dataservice_test.go [hunk 6]
```diff
@@ -627,6 +628,88 @@ var _ = Describe("{UpgradeDataServiceImage}", func() {
...
+					log.InfoD("Pods found on %v node. Trying to Cordon and Drain pods from this node now.", nodeName)
+					err = k8sCore.DrainPodsFromNode(nodeName, podsList, timeOut, maxtimeInterval)
+					log.FailOnError(err, "draining pod from the node failed")
+				})
```
Comment (lsrinivas-pure) [inline tests/pds/dataservice_test.go:663]: same here

File: tests/pds/dataservice_test.go [hunk 7]
```diff
@@ -627,6 +628,88 @@ var _ = Describe("{UpgradeDataServiceImage}", func() {
...
+				Step("Validate Data Service Again", func() {
+					resourceTemp, storageOp, config, err := pdslib.ValidateDataServiceVolumes(deployment, ds.Name, dataServiceDefaultResourceTemplateID, storageTemplateID, namespace)
+					log.FailOnError(err, "error on ValidateDataServiceVolumes method")
```
Comment (lsrinivas-pure) [inline tests/pds/dataservice_test.go:666]: add service name in the log

File: tests/pds/dataservice_test.go [hunk 8]
```diff
@@ -627,6 +628,88 @@ var _ = Describe("{UpgradeDataServiceImage}", func() {
...
+					ValidateDeployments(resourceTemp, storageOp, config, ds.Replicas, dataServiceVersionBuildMap)
+				})
+				Step("Validate no pods are on the cordoned node anymore", func() {
+					nodes, err := pdslib.GetNodesOfSS(*deployment.ClusterResourceName, namespace)
```
Comment (lsrinivas-pure) [inline tests/pds/dataservice_test.go:670]: add service name in the log

File: tests/pds/dataservice_test.go [hunk 9]
```diff
@@ -627,6 +628,88 @@ var _ = Describe("{UpgradeDataServiceImage}", func() {
...
+					log.FailOnError(err, "Cannot fetch nodes of the running Data Service")
+					for _, nodeObj := range nodes {
+						log.InfoD("Node Name is: %s", nodeObj.Name)
+						if nodeObj.Name == nodeName {
```
Comment (lsrinivas-pure) [inline tests/pds/dataservice_test.go:674]: do u need this data to be pushed to dashboard?

File: tests/pds/dataservice_test.go [hunk 10]
```diff
@@ -627,6 +628,88 @@ var _ = Describe("{UpgradeDataServiceImage}", func() {
...
+							log.FailOnError(errors.New("New Pod came up on the node that was cordoned."), "Unexpected error")
+						}
+					}
+					log.InfoD("The pods of the Stateful Set %v are not on the cordoned node. Moving ahead now.", *deployment.ClusterResourceName)
+				})
+				Step("Running Workloads before scaling up of dataservices ", func() {
+					log.InfoD("Running Workloads on DataService %v ", ds.Name)
+					var params pdslib.WorkloadGenerationParams
+					pod, dep, err = RunWorkloads(params, ds, deployment, namespace)
+					log.FailOnError(err, fmt.Sprintf("Error while genearating workloads for dataservice [%s]", ds.Name))
```
Comment (lsrinivas-pure) [inline tests/pds/dataservice_test.go:684]: nit: generating

File: tests/pds/dataservice_test.go [hunk 11]
```diff
@@ -627,6 +628,88 @@ var _ = Describe("{UpgradeDataServiceImage}", func() {
...
+					}
+					log.FailOnError(err, "error deleting workload generating pods")
+				})
+			}
+		})
+	})
+	JustAfterEach(func() {
+		defer EndTorpedoTest()
+
+		defer func() {
```
Comment (lsrinivas-pure) [inline tests/pds/dataservice_test.go:702]: no need to defer this function as this is executing last

File: tests/pds/dataservice_test.go [hunk 12]
```diff
@@ -627,6 +628,88 @@ var _ = Describe("{UpgradeDataServiceImage}", func() {
...
+			if !isDeploymentsDeleted {
+				Step("Delete created deployments")
+				resp, err := pdslib.DeleteDeployment(deployment.GetId())
+				log.FailOnError(err, "Error while deleting data services")
+				dash.VerifyFatal(resp.StatusCode, http.StatusAccepted, "validating the status response")
```
Comment (lsrinivas-pure) [inline tests/pds/dataservice_test.go:707]: add deployment id in the error log

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/pds/dataservice_test.go (hunk 6 above)
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go:653]: Done

File: tests/pds/dataservice_test.go (hunk 8 above)
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go:660]: Done

File: tests/pds/dataservice_test.go (hunk 9 above)
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go:663]: Thanks - removed this log line

File: tests/pds/dataservice_test.go (hunk 11 above)
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go:691]: Noted. This needs to be changed in every TC and not just this one. Will raise a separate PR for this one later that will take care of all TCs

File: tests/pds/dataservice_test.go (hunk 12 above)
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go:696]: Noted. This needs to be changed in every TC and not just this one. Will raise a separate PR for this one later that will take care of all TCs

File: drivers/pds/lib/utils.go (hunk 1 above)
Comment (dbhatnagar-px) [reply drivers/pds/lib/utils.go]: Done

File: tests/pds/dataservice_test.go (hunk 3 above)
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go]: Done

File: tests/pds/dataservice_test.go (hunk 4 above)
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go]: Thanks - Removed this log line - was there just for testing purposes

File: tests/pds/dataservice_test.go (hunk 10 above)
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go]: Done

File: tests/pds/dataservice_test.go (hunk 5 above)
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go]: Done

File: tests/pds/dataservice_test.go (hunk 7 above)
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go]: Done

**Review by Madanagopal19 — APPROVED** [review]
LGTM

**Review by gauraa2 — APPROVED** [review]
LGTM

**Pull request merged**